src/device_manager.rs

```diff
@@ -0,0 +1,360 @@
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    if self
+                        .pio_bus
+                        .insert(IoRange::new_pio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    if self
+                        .mmio_bus
+                        .insert(IoRange::new_mmio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    /// Unregister a device from `IoManager`, e.g. users specified removing.
+    /// VMM pre-fetches the resources e.g. dev.get_assigned_resources()
+    /// VMM is responsible for freeing the resources.
+    ///
+    /// # Arguements
+    ///
+    /// * `resources`: resources that this device owns, might include
+    ///                port I/O and memory-mapped I/O ranges, irq number, etc.
+    pub fn unregister_device_io(&mut self, resources: &[Resource]) -> Result<()> {
+        for res in resources.iter() {
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    self.pio_bus.remove(&IoRange::new_pio_range(base, size));
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    self.mmio_bus.remove(&IoRange::new_mmio_range(base, size));
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    // BTreeMap::range() yields elements from min(inclusive) to max(exclusive).
+    // Let's use a fake range for getting inclusive result.
+    fn get_entry(&self, addr: IoAddress) -> Option<(&IoRange, &Arc<dyn DeviceIo>)> {
+        match addr {
+            IoAddress::Pio(a) => self
+                .pio_bus
+                .range(..&IoRange::new_pio_range(a + 1, 0))
+                .nth_back(0),
+            IoAddress::Mmio(GuestAddress(a)) => self
+                .mmio_bus
+                .range(..&IoRange::new_mmio_range(a + 1, 0))
+                .nth_back(0),
+        }
+    }
+
+    // Return the Device mapped the address.
+    fn get_device(&self, addr: IoAddress) -> Option<Arc<dyn DeviceIo>> {
+        if let Some((range, dev)) = self.get_entry(addr) {
+            if (addr.raw_value() - range.base.raw_value()) < range.size.raw_value() {
+                return Some(dev.clone());
```

Comment:
Arc::clone() is still a relative heavy operations and DeviceIo::{read, write} takes an &self. So could we return &Arc<dyn DeviceIo> instead of Arc<dyn DeviceIo> here?